L80-csharp-load-balancing.md

```diff
@@ -0,0 +1,81 @@
+Title
+----
+* Author(s): James Newton-King
+* Approver: a11r
+* Status: Draft
+* Implemented in: csharp
+* Last updated: 2021-05-19
+* Discussion at: <google group thread> (filled after thread exists)
+
+## Abstract
+
+grpc-dotnet currently has no load balancing support. A channel is always created for a single address.
+
+This documentation outlines an initial load balancing implementation for the grpc-dotnet client.
+
+## Background
+
+grpc-dotnet currently doesn't support any form of load balancing in the client. The changes outlined in this proposal will enable basic gRPC load balancing and some of its related features. These include:
+
+* Name resolving. Built-in support is included for resolving from DNS, and a static resolver. Custom resolvers can be created.
+* Load balancing strategies. Built-in support is included for pick first and round robin load balancing. Custom load balancers can be created.
+* Connectivity status at sub-channel and channel level
+* Wait for ready
+
+## Proposal
+
+![Diagram](L80_graphics/diagram.png)
+
+### Initialization
+
+A channel is created with an address, an optional service provider, and an optional service config.
+
+```csharp
+var channel = GrpcChannel.ForAddress(
+    "dns://backend.default.svc.cluster.local",
+    new GrpcChannelOptions { ChannelCredentials = ChannelCredentials.Insecure });
+var client = new Greet.GreeterClient(channel);
+
+var response = await client.SayHelloAsync(new HelloRequest { Name = "world" });
+Console.WriteLine(response.Message);
+```
+
+* Address scheme will be used to create a resolver.
+  * Existing logic allows for `http` and `https`. They are special-cased to never use a resolver.
+  * A collection of resolver factories are fetched from the service provider. One is selected using the address scheme and it will create resolver. For example, `dns` will match to `DnsAddressResolverFactory` which will create a `DnsAddressResolver`.
+* Load balancer name in service configure will be used to create a load balancer.
+  * A collection of load balancer factories are fetched from the service provider. One is selected using the configured name (if none specified then `pick_first` is used) and it will create load balancer.
+  * Load balancer is initialized with an `IChannelController` instance by the channel. This is used to call back to the channel to create sub-channels and update the picker.
+
+### Connecting
+
+* No network activity until `GrpcChannel.ConnectAsync` is called, or the first RPC is started.
+* On connect, the `ConnectionManager` will subscribe to address resolver and call the resolver's `RefreshAsync` to prompt the resolver to run immediately.
+* Resolver will push addresses to `ConnectionManager` which will then call balancer.
+* Balancer will use `IChannelController` to create, remove or update sub-channels as needed based on the addresses.
```

Comment:
This is the type:
```csharp
    public interface IConnectionController {
        ConnectivityState State { get; }
        SubChannel CreateSubChannel(SubChannelOptions options);
        void RemoveSubChannel(SubChannel subChannel);
        Task ResolveNowAsync(CancellationToken cancellationToken);
        void UpdateAddresses(SubChannel subChannel, IReadOnlyList<DnsEndPoint> addresses);
        void UpdateState(BalancerState state);
    }
```

So:
* Move `UpdateAddresses` onto `SubChannel`.
* Delete `RemoveSubChannel`. Right now calling this method signals that the sub-channel should be removed form the pool. Without `RemoveSubChannel` is calling shutdown on a sub-channel the signal that it should be removed from the pool?

Reply:
I think I see now what implementations mean when they refer to a subchannel pool. It allows subchannels to be shared between load balancers.

If that is the case then grpc-dotnet doesn't have a subchannel pool. The channel does keep a list of non-shutdown subchannels (this is what I thought a subchannel pool referred to), however, these aren't shared between load balancers.

At the moment this seems fine, but I believe xDS will have situations where there will be multiple load balancers. I think I need to look at what Java and Go do in the situation of multiple load balancers.

> The real subchannel will have a lot of methods that are needed for internal use but that do not need to be exposed to the LB policy.

.NET allows for internal methods. Exposing the type should be fine. And if it ever isn't fine in the future then adding a new internal type won't be a breaking change.

Reply:
> This is the type:
> 
> ```cs
>     public interface IConnectionController {
>         ConnectivityState State { get; }
>         SubChannel CreateSubChannel(SubChannelOptions options);
>         void RemoveSubChannel(SubChannel subChannel);
>         Task ResolveNowAsync(CancellationToken cancellationToken);
>         void UpdateAddresses(SubChannel subChannel, IReadOnlyList<DnsEndPoint> addresses);
>         void UpdateState(BalancerState state);
>     }
> ```
> 
> So:
> 
> * Move `UpdateAddresses` onto `SubChannel`.
> * Delete `RemoveSubChannel`. Right now calling this method signals that the sub-channel should be removed form the pool. Without `RemoveSubChannel` is calling shutdown on a sub-channel the signal that it should be removed from the pool?

I looked back at grpc-go and its version of this type [has RemoveSubChannel and UpdateAddresses](https://github.com/grpc/grpc-go/blob/3dd75a6888ce5d1b5195c5cf72241d9e36f68e42/balancer/balancer.go#L146-L155). Also, UpdateAddresses was originally on the SubChannel [and was moved](https://github.com/grpc/grpc-go/blob/3dd75a6888ce5d1b5195c5cf72241d9e36f68e42/balancer/balancer.go#L97-L107). It would be good to understand why golang has this API from one of its maintainers.

Reply:
> At the moment this seems fine, but I believe xDS will have situations where there will be multiple load balancers. I think I need to look at what Java and Go do in the situation of multiple load balancers.

In general, this is something that Java and Go handle much less gracefully than C-core does.  They usually wind up creating duplicate subchannels in this case.  Although to be fair, this generally comes up only when a configuration change occurs, which is not that often, so it's hard to know if it will cause problems in practice.

We've talked about ways to ameliorate this if it becomes a problem, such as possibly introducing an LB policy higher in the tree to provide subchannel reuse for LB policies underneath it.  That would effectively result in the same behavior as the C-core subchannel pool, except it would be implemented in an LB policy instead of by the channel itself.

> If that is the case then grpc-dotnet doesn't have a subchannel pool. The channel does keep a list of non-shutdown subchannels (this is what I thought a subchannel pool referred to), however, these aren't shared between load balancers.

Okay, then please update the diagram and the text to reflect that there is no subchannel pool in this design.

> I looked back at grpc-go and its version of this type [has RemoveSubChannel and UpdateAddresses](https://github.com/grpc/grpc-go/blob/3dd75a6888ce5d1b5195c5cf72241d9e36f68e42/balancer/balancer.go#L146-L155). Also, UpdateAddresses was originally on the SubChannel [and was moved](https://github.com/grpc/grpc-go/blob/3dd75a6888ce5d1b5195c5cf72241d9e36f68e42/balancer/balancer.go#L97-L107). It would be good to understand why golang has this API from one of its maintainers.

@dfawley or @menghanl may be able to explain the history here.